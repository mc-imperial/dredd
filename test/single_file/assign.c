int main() {
  int x;
  volatile int y;
  x = 2;
  y = 4;
}
