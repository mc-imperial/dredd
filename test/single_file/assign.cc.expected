#include <cinttypes>
#include <cstddef>
#include <functional>
#include <string>


#ifdef _MSC_VER
#define thread_local __declspec(thread)
#elif __APPLE__
#define thread_local __thread
#endif

static thread_local bool __dredd_some_mutation_enabled = true;
static bool __dredd_enabled_mutation(int local_mutation_id) {
  static thread_local bool initialized = false;
  static thread_local uint64_t enabled_bitset[1];
  if (!initialized) {
    bool some_mutation_enabled = false;
    const char* dredd_environment_variable = std::getenv("DREDD_ENABLED_MUTATION");
    if (dredd_environment_variable != nullptr) {
      std::string contents(dredd_environment_variable);
      while (true) {
        size_t pos = contents.find(",");
        std::string token = (pos == std::string::npos ? contents : contents.substr(0, pos));
        if (!token.empty()) {
          int value = std::stoi(token);
          int local_value = value - 0;
          if (local_value >= 0 && local_value < 32) {
            enabled_bitset[local_value / 64] |= (static_cast<uint64_t>(1) << (local_value % 64));
            some_mutation_enabled = true;
          }
        }
        if (pos == std::string::npos) {
          break;
        }
        contents.erase(0, pos + 1);
      }
    }
    initialized = true;
    __dredd_some_mutation_enabled = some_mutation_enabled;
  }
  return (enabled_bitset[local_mutation_id / 64] & (static_cast<uint64_t>(1) << (local_mutation_id % 64))) != 0;
}

#define MUTATION_RETURN(arg) arg
#define REPLACE_EXPR_NOT(args, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return args
#define REPLACE_EXPR_MINUS(args, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return args
#define REPLACE_EXPR_INT_ZERO(args, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return args
#define REPLACE_EXPR_INT_ONE(args, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return args
#define REPLACE_EXPR_INT_MINUS_ONE(args, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return args
#define REPLACE_BINARY_XorAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 ^= arg2;
#define REPLACE_BINARY_XorAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 ^= arg2;
#define REPLACE_BINARY_SubAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 -= arg2;
#define REPLACE_BINARY_SubAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 -= arg2;
#define REPLACE_BINARY_ShrAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 >>= arg2;
#define REPLACE_BINARY_ShrAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 >>= arg2;
#define REPLACE_BINARY_ShlAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 <<= arg2;
#define REPLACE_BINARY_ShlAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 <<= arg2;
#define REPLACE_BINARY_RemAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 %= arg2;
#define REPLACE_BINARY_RemAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 %= arg2;
#define REPLACE_BINARY_OrAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 |= arg2;
#define REPLACE_BINARY_OrAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 |= arg2;
#define REPLACE_BINARY_MulAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 *= arg2;
#define REPLACE_BINARY_MulAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 *= arg2;
#define REPLACE_BINARY_DivAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 /= arg2;
#define REPLACE_BINARY_DivAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 /= arg2;
#define REPLACE_BINARY_AndAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 &= arg2;
#define REPLACE_BINARY_AndAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 &= arg2;
#define REPLACE_BINARY_AddAssign_arg1_volatile_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 += arg2;
#define REPLACE_BINARY_AddAssign_arg1_int_arg2_int(arg1, arg2, mutation_id_offset) if (__dredd_enabled_mutation(local_mutation_id + mutation_id_offset)) return arg1 += arg2;
#define MUTATION_PRELUDE(arg) if (!__dredd_some_mutation_enabled) return arg
static volatile int& __dredd_replace_binary_operator_Assign_arg1_volatile_int_arg2_int(volatile int& arg1, int arg2, int local_mutation_id) {
  MUTATION_PRELUDE(arg1 = arg2);
  REPLACE_BINARY_AddAssign_arg1_volatile_int_arg2_int(arg1, arg2, 0);
  REPLACE_BINARY_AndAssign_arg1_volatile_int_arg2_int(arg1, arg2, 1);
  REPLACE_BINARY_DivAssign_arg1_volatile_int_arg2_int(arg1, arg2, 2);
  REPLACE_BINARY_MulAssign_arg1_volatile_int_arg2_int(arg1, arg2, 3);
  REPLACE_BINARY_OrAssign_arg1_volatile_int_arg2_int(arg1, arg2, 4);
  REPLACE_BINARY_RemAssign_arg1_volatile_int_arg2_int(arg1, arg2, 5);
  REPLACE_BINARY_ShlAssign_arg1_volatile_int_arg2_int(arg1, arg2, 6);
  REPLACE_BINARY_ShrAssign_arg1_volatile_int_arg2_int(arg1, arg2, 7);
  REPLACE_BINARY_SubAssign_arg1_volatile_int_arg2_int(arg1, arg2, 8);
  REPLACE_BINARY_XorAssign_arg1_volatile_int_arg2_int(arg1, arg2, 9);
  return MUTATION_RETURN(arg1 = arg2);
}

static int& __dredd_replace_binary_operator_Assign_arg1_int_arg2_int(int& arg1, int arg2, int local_mutation_id) {
  MUTATION_PRELUDE(arg1 = arg2);
  REPLACE_BINARY_AddAssign_arg1_int_arg2_int(arg1, arg2, 0);
  REPLACE_BINARY_AndAssign_arg1_int_arg2_int(arg1, arg2, 1);
  REPLACE_BINARY_DivAssign_arg1_int_arg2_int(arg1, arg2, 2);
  REPLACE_BINARY_MulAssign_arg1_int_arg2_int(arg1, arg2, 3);
  REPLACE_BINARY_OrAssign_arg1_int_arg2_int(arg1, arg2, 4);
  REPLACE_BINARY_RemAssign_arg1_int_arg2_int(arg1, arg2, 5);
  REPLACE_BINARY_ShlAssign_arg1_int_arg2_int(arg1, arg2, 6);
  REPLACE_BINARY_ShrAssign_arg1_int_arg2_int(arg1, arg2, 7);
  REPLACE_BINARY_SubAssign_arg1_int_arg2_int(arg1, arg2, 8);
  REPLACE_BINARY_XorAssign_arg1_int_arg2_int(arg1, arg2, 9);
  return MUTATION_RETURN(arg1 = arg2);
}

static int __dredd_replace_expr_int_constant(int arg, int local_mutation_id) {
  MUTATION_PRELUDE(arg);
  REPLACE_EXPR_NOT(arg, 0);
  REPLACE_EXPR_MINUS(arg, 1);
  REPLACE_EXPR_INT_ZERO(0, 2);
  REPLACE_EXPR_INT_ONE(1, 3);
  REPLACE_EXPR_INT_MINUS_ONE(-1, 4);
  return MUTATION_RETURN(arg);
}

int main() {
  int x;
  volatile int y;
  if (!__dredd_enabled_mutation(15)) { __dredd_replace_binary_operator_Assign_arg1_int_arg2_int(x , __dredd_replace_expr_int_constant(2, 0), 5); }
  if (!__dredd_enabled_mutation(31)) { __dredd_replace_binary_operator_Assign_arg1_volatile_int_arg2_int(y , __dredd_replace_expr_int_constant(4, 16), 21); }
}
